cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -207,8 +215,10 @@ protected void run() {
         Thread.currentThread().interrupt();
       }
     }
-    // Persist records and messageId's after all ProcessMetricsThread's complete.
-    persistRecordsMessageIds(records, topicMessageIds);
+    // Persist metricsFromAllTopics and messageId's after all ProcessMetricsThread's complete.
+    if (!metricsFromAllTopics.isEmpty()) {
```

Comment:
Here all threads already complete and no thread is writing to the queue. Is it still necessary to make a copy?